keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,705 @@
+cloud provider, and it may return multiple IPs of each type (eg, if
+the node has multiple network interfaces, or multiple addresses on a
+single interface). eg:
+
+    [ { "type": "ExternalIP", "address": "42.42.42.42" },
+      { "type": "InternalIP", "address": "10.1.2.3" },
+      { "type": "InternalIP", "address": "10.4.5.6" },
+      { "type": "InternalIP", "address": "10.7.8.9" } ]
+
+In this case, the primary IP would be `10.1.2.3`. If the administrator
+wants a different primary IP, they can pass `--node-ip` to choose
+another one, but that will have the side effect of removing all other
+IPs of the same type from the list. Eg, passing `--node-ip 10.4.5.6`,
+would result in `Node.Status.Addresses` having the value:
+
+    [ { "type": "ExternalIP", "address": "42.42.42.42" },
+      { "type": "InternalIP", "address": "10.4.5.6" } ]
+
+When using an external cloud-provider, the administrator has no
+control over the node IPs; in the above example, passing `--node-ip
+10.4.5.6` would have no effect on the returned list of addresses, and
+the primary IP would still be `10.1.2.3`. (Although passing `--node-ip
+10.11.12.13` would result in an error since that doesn't appear in the
+list at all.)
+
+#### API Cleanliness
+
+`Node.Status.Addresses` is annotated with an incorrect strategic-merge
+patch annotation, and so it is not possible to use strategic merge
+patch to change that field without losing information. (See discussion
+in [kubernetes #79391].) Although kubelet and cloud-controller-manager
+have been fixed to work around the problem, it is still annoying to
+have the incorrect annotation there, and it is possible that other
+components might attempt to update the field incorrectly in the
+future.
+
+Additionally, `Node.Status.Addresses` is not currently sufficiently
+checked by validation; it is possible for an external cloud provider
+to declare an `InternalIP` or `ExternalIP` address that is not a
+syntactically-valid IP address, meaning clients that read this field
+may need to sanity-check it themselves ([kubernetes #89817]). Adding
+validation now could break deployed cloud providers (which might
+otherwise appear to be fully-functional, as long as the primary IP was
+valid).
+
+[kubernetes #79391]: https://github.com/kubernetes/kubernetes/pull/79391
+[kubernetes #89817]: https://github.com/kubernetes/kubernetes/pull/89817
+
+#### `NodeAddressType` Semantics
+
+The difference between `InternalIP` and `ExternalIP` node addresses is
+not well-defined except in the case of IPv4 addresses in public
+clouds, and the two types are used inconsistently outside of that
+environment. (eg, See [kubernetes #86918] as an example of uncertainty
+about the right type to use in a particular situation.)
+
+[kubernetes #86918]: https://github.com/kubernetes/kubernetes/pull/86918#discussion_r363997750
+
+### Goals
+
+- Deprecate `Node.Status.Addresses` and replace it with new fields
```

Comment:
Removing the v1 field is not allowed until v2 Node exists, and v1 Node is deprecated and removed (none of which is planned). What this would actually look like is adding a second field with duplicate information, and reasoning through how not to break existing readers/writers to the old field.

Reply:
> how not to break existing readers/writers to the old field

FTR I don't think this is explicit anywhere but I'm assuming we don't need to care about any writers other than cloud-controller-manager and kubelet, because both of them assume they have sole ownership over the field, and if anyone else tries to write to it, kubelet/ccm will quickly overwrite their changes.